kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs

```diff
@@ -562,32 +577,38 @@ impl<'tcx> GotocCtx<'tcx> {
                     _ => unreachable!("must be array or slice"),
                 }
             }
-            ProjectionElem::Downcast(_, idx) => {
+            ProjectionElem::Downcast(idx) => {
                 // downcast converts a variable of an enum type to one of its discriminated cases
-                let t = before.mir_typ();
-                let (case_name, type_or_variant) = match t.kind() {
-                    ty::Adt(def, _) => {
-                        let variant = def.variant(idx);
-                        (variant.name.as_str().into(), TypeOrVariant::Variant(variant))
+                let ty = before.mir_typ();
+                let ty_kind = ty.kind();
+                let (case_name, type_or_variant) = match &ty_kind {
+                    TyKind::RigidTy(RigidTy::Adt(def, _)) => {
+                        let variant = def.variant(*idx).unwrap();
+                        (variant.name().into(), TypeOrVariant::Variant(variant))
                     }
-                    ty::Coroutine(..) => {
-                        (self.coroutine_variant_name(idx), TypeOrVariant::CoroutineVariant(idx))
+                    TyKind::RigidTy(RigidTy::Coroutine(..)) => {
+                        let idx_internal = rustc_internal::internal(idx);
+                        (
+                            self.coroutine_variant_name(idx_internal),
+                            TypeOrVariant::CoroutineVariant(*idx),
+                        )
                     }
                     _ => unreachable!(
                         "cannot downcast {:?} to a variant (only enums and coroutines can)",
-                        &t.kind()
+                        &ty.kind()
                     ),
                 };
-                let layout = self.layout_of(t);
+                let layout = self.layout_of(rustc_internal::internal(ty));
                 let expr = match &layout.variants {
                     Variants::Single { .. } => before.goto_expr,
                     Variants::Multiple { tag_encoding, .. } => match tag_encoding {
                         TagEncoding::Direct => {
-                            let cases = if t.is_coroutine() {
```

Comment:
Let's have `is_coroutine` like we do `is_box`.

Reply:
Is it OK if I create an util method for now in Kani. StableMIR doesn't have all the is_<type> methods yet, and I don't think its high priority right now, since users can still check that, like we do.

Reply:
Of course, you can add it next to `is_box`.